### PR TITLE
OAuth support for the Swift SDK

### PR DESCRIPTION
**Pull request by jkaster**
Added the `OAuthSession` class to support OAuth flows

### CODE REVIEW

**Comment by jkaster** [comment]
Thanks for the review, Danny. I think we'll leave this in a branch until `refresh_token` is available in the next Looker release.

File: kotlin/src/test/TestAuthSession.kt [hunk 1]
```diff
@@ -88,9 +88,12 @@ class TestAuthSession {
     @test
     fun testRedemptionBody() {
         val session = OAuthSession(config.oAuthTestSettings, Transport(testSettings))
-        val request = session.redeemAuthCodeBody("authCode", "com.looker.android")
-        assertEquals(request["code"], "authCode")
-        assertEquals(request["code_verifier"], "636f6d2e6c6f6f6b65722e616e64726f6964")
+        val hashCode = session.sha256hash("com.looker.android")
```
[the remark this answers is not in the document]
Comment (jkaster) [reply kotlin/src/test/TestAuthSession.kt]: I think that was Danny's intent that the hashed value is passed in.

File: swift/looker/rtl/oauthSession.swift [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+    override init(_ settings: IApiSettings, _ transport: ITransport? = nil) {
+        super.init(settings, transport)
+    }
+
+    func requestToken(_ body: Values) throws -> AuthToken {
+        let response : SDKResponse<AccessToken, SDKError> = self.transport.request(
+            HttpMethod.POST,
+            "/api/token",
+            nil,
+            body,
```
Comment (rlm1023) [inline swift/looker/rtl/oauthSession.swift]: You probably need to encode these params yeah? I found this request would 500 on the server side unless this was changed to:
```suggestion
            encodeParams(body),
```

Comment (jkaster) [reply swift/looker/rtl/oauthSession.swift]: Thanks. Fixed.

File: kotlin/src/main/com/looker/rtl/OAuthSession.kt [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+        return this.authToken
+    }
+
+    override fun getToken(): AuthToken {
+        if (!this.isAuthenticated()) {
+            if (this.activeToken().refreshToken?.isNotEmpty()!!) {
+                val config = this.apiSettings.readConfig()
+                // fetch the token
+                this.requestToken(mapOf(
+                        "grant_type" to "request_token",
```
Comment (rlm1023) [inline kotlin/src/main/com/looker/rtl/OAuthSession.kt]: This should be `refresh_token` not `request_token`.

File: swift/looker/rtl/authToken.swift [hunk 4]
```diff
@@ -37,7 +37,7 @@ protocol AccessTokenProtocol {
 //  var expires_in: Int
 //}
 
-struct AuthToken: AccessTokenProtocol {
+struct AuthToken: AccessTokenProtocol, Codable {
```
Comment (jkaster) [inline swift/looker/rtl/authToken.swift]: Thanks Rob

**Comment by jkaster** [comment]
@joeldodge79 can you verify Python tests are behaving as you expect?

File: kotlin/src/main/com/looker/rtl/OAuthSession.kt [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+        val lookerUrl = config["looker_url"]
+        return addQueryParams("$lookerUrl/auth", mapOf(
+                "response_type" to "code",
+                "client_id" to config["client_id"],
+                "redirect_uri" to config["redirect_uri"],
+                "scope" to scope,
+                "state" to state,
+                "code_challenge_method" to "S256",
+                "code_challenge" to this.codeVerifier
+        ))
```
Comment (rlm1023) [inline kotlin/src/main/com/looker/rtl/OAuthSession.kt:156]: This isn't quite right. The code verifier is the raw secure random number. The code challenge is that number hashed.

```suggestion
        this.codeVerifier = this.secureRandom(32)
        val codeChallenge = this.sha256hash(this.codeVerifier)
        val config = this.apiSettings.readConfig()
        val lookerUrl = config["looker_url"]
        return addQueryParams("$lookerUrl/auth", mapOf(
                "response_type" to "code",
                "client_id" to config["client_id"],
                "redirect_uri" to config["redirect_uri"],
                "scope" to scope,
                "state" to state,
                "code_challenge_method" to "S256",
                "code_challenge" to codeChallenge
        ))
```

File: swift/looker/rtl/oauthSession.swift [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+            [
+                "response_type": "code",
+             "client_id": config["client_id"],
+            "redirect_uri": config["redirect_uri"],
+            "scope": scope,
+            "state": state,
+            "code_challenge_method": "S256",
+            "code_challenge": self.code_verifier
+        ])
+        return url
```
Comment (rlm1023) [inline swift/looker/rtl/oauthSession.swift:117]: Looks like same issue here:
```suggestion
        self.code_verifier = try! self.secureRandom(32)
        let code_challenge = self.sha256Hash(self.code_verifier)
        let config = self.settings.readConfig(nil)
        let looker_url = config["looker_url"]!
        let url = addQueryParams("\(looker_url)/auth", [
            "response_type": "code",
            "client_id": config["client_id"],
            "redirect_uri": config["redirect_uri"],
            "scope": scope,
            "state": state,
            "code_challenge_method": "S256",
            "code_challenge": code_challenge
        ])
        return url
```

**Comment by rlm1023** [comment]
Oh those might not be that simple tho because types...

File: swift/looker/rtl/oauthSession.swift [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+            }
+        }
+        return self.activeToken
+    }
+
+    /*
+     Generate an OAuth2 authCode request URL
+     */
+    func createAuthCodeRequestUrl(scope: String, state: String) throws -> String {
+        self.code_verifier = try! self.secureRandom(32)
```
Comment (rlm1023) [inline swift/looker/rtl/oauthSession.swift:104]:
```suggestion
        self.code_verifier = try! self.secureRandom(32).hexStr()
```

File: kotlin/src/main/com/looker/rtl/OAuthSession.kt [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+            }
+        }
+        return this.activeToken()
+    }
+
+    /**
+     * Generate an OAuth2 authCode request URL
+     */
+    fun createAuthCodeRequestUrl(scope: String, state: String): String {
+        this.codeVerifier = this.secureRandom(32)
```
[the remark this answers is not in the document]
Comment (jkaster) [reply kotlin/src/main/com/looker/rtl/OAuthSession.kt:144]: Heh. This is one of the things I was going to ask you about!

**Pull request merged**